README.rst
collective.eggproxy package
===========================

.. contents::

What is collective.eggproxy ?
-----------------------------

collective.eggproxy is a smart mirror for PyPI.

It will collect packages on PyPI only when a program like easy_install
or zc.buildout asks for it. In other words, unlike some mirrors
that act like rsync and get the whole PyPI base (more than 5 gigas)
collective.eggproxy will only get what you need.

At first run collective.eggproxy downloads pypi index and builds a page of links.
When a software asks for a specific package, version, etc.
collective.eggproxy downloads it if needed and store it locally.

How to use collective.eggproxy ?
--------------------------------

After it has been installed, you can just launch it as a standalone server
like this::

    $ mkdir /tmp/eggs
    $ eggproxy_run

The proxy will then be available on the localhost on the port 8888.
All package will be downloaded by default into /var/www. If this directory
does not exists (or if you are under windows), you will need to configure it,
as explained in the next section.

From there you can use it in easy_install like this::

    easy_install -i http://localhost:8888/ -H "*localhost*" iw.fss

The iw.fss package will be downloaded, stored locally and provided to easy_insall.

In zc.buildout, just define the index value in the buildout section::

    [buildout]

    index = http://localhost:8888/
    allow-hosts = *localhost*

That's it !

Advanced configuration
----------------------

collective.eggproxy can use a configuration file like this::

    [eggproxy]
    eggs_directory = /path/to/our/cache
    index = http://pypi.python.org/simple

    # Update information for files older than 24h
    update_interval = 24
    # Port number where proxy will run
    port = 8888
    # always_refresh is off by default. Setting it to 1 forces eggproxy to
    # always attempt a pypi connection. Slower but fresher.
    always_refresh = 0
    # timeout is only needed for always_refresh: it sets a socket timeout.
    timeout = 3

All options apart from eggs_directory are optional.

collective.eggproxy looks in three locations:

* An ``eggproxy.conf`` in your homedir. This can also be convenient on Windows
  where ``/etc/`` doesn't exist.

* ``../etc/eggproxy.conf`` as seen from the binary, which you can use for
  buildout setups (see buildout documentation below).

* ``/etc/eggproxy.conf``.


Running the proxy using Paste
-----------------------------

You need a paste configuration file::

  $ paster make-config collective.eggproxy myconfig.ini

Feel free to edit the default configuration.
This configuration will override the ``eggproxy.conf`` file.

Then use ``paster`` to serve the application::

  $ paster serve configfile.ini

And test it::

  $ easy_install -i http://localhost:8888/ -H "*localhost*" iw.fss


Installing collective.eggproxy in a buildout
--------------------------------------------

A quick way to set up collective.eggproxy is by installing it in a
buildout.  The advantage is that it is all nicely contained in one
directory.  You can use the following buildout config as an example::

  [buildout]
  parts = console_scripts configuration

  [console_scripts]
  recipe = zc.recipe.egg
  eggs = collective.eggproxy

  [configuration]
  recipe = collective.recipe.template
  input = etc/eggproxy.conf.in
  output = etc/eggproxy.conf


This will put the eggproxy_run and eggproxy_update scripts in the ``bin/`` directory.
Add a subdirectory ``etc/`` inside the buildout with an ``eggproxy.conf.in``
file::

  [eggproxy]
  eggs_directory = ${buildout:directory}/var/cache
  #update_interval = 24
  #index = http://pypi.python.org/simple
  #port = 8888

`collective.recipe.template
<http://pypi.python.org/pypi/collective.recipe.template>`_ will turn that into
an etc/eggproxy.conf with the correct settings


Using the proxy behind Apache
-----------------------------

You can also use collective.eggproxy with Apache. You will need for that
mod_python 3.3 for apache. It will not work with previous versions.

Debian Etch users: Etch provides 3.2, so users of Etch must get the source
package from "Lenny" (testing) and rebuild it with dpkg-buildpackage. Please have
a look at the related section in this document to get some help.

An Apache setup for http://servername/pypi can be::

    Alias /pypi "/path/to/pypi"
    <Directory "/path/to/pypi">
        Order allow,deny
        Allow from all
        SetHandler mod_python
        PythonFixupHandler collective.eggproxy.apache_handler::fixup_handler
        PythonInterpreter main_interpreter
        PythonOption URLPrefix /pypi
        PythonPath "sys.path+['/path/to/collective.eggproxy']"
    </Directory>

Apache must be able to write to "/path/to/pypi"! (usually Apache user is
``www-data``). For updating proxied egg information you must add the script
``eggproxy_update`` into the crontab of this user.

mod_python update on Debian Etch
::::::::::::::::::::::::::::::::

Quick build instructions from debian "lenny" sources::

    $ sudo apt-get build-dep libapache2-mod-python
    $ sudo apt-get install fakeroot

Visit http://packages.debian.org/sources/lenny/libapache2-mod-python and grab
the files with these extensions:

- .dsc
- .orig.tar.gz
- .diff.gz

::

    $ wget http://ftp.de.debian.org/debian/pool/main/liba/libapache2-mod-python/libapache2-mod-python_3.3.1-3.dsc
    $ wget http://ftp.de.debian.org/debian/pool/main/liba/libapache2-mod-python/libapache2-mod-python_3.3.1.orig.tar.gz
    $ wget http://ftp.de.debian.org/debian/pool/main/liba/libapache2-mod-python/libapache2-mod-python_3.3.1-3.diff.gz
    $ tar xpvzf libapache2-mod-python_3.3.1.orig.tar.gz
    $ cd mod_python-3.3.1/
    $ gzip -dc ../libapache2-mod-python_3.3.1-3.diff.gz |patch -p1
    $ chmod +x debian/rules
    $ dpkg-buildpackage -rfakeroot -b
    $ cd ..

you should have the following files:

- libapache2-mod-python_3.3.1-3_i386.deb
- libapache2-mod-python-doc_3.3.1-3_all.deb

The last step is calling ``dpkg``::

    $ sudo dpkg -i libapache2-mod-python_3.3.1-3_i386.deb


Development
-----------

Development happens now on github at
https://github.com/bbinet/collective.eggproxy

The issue tracker is there, too. Feel free to make bug reports there or to
fork the code to fix errors. That's what github is The.

(Historic note: the code used to live at
https://svn.plone.org/svn/collective/collective.eggproxy/trunk)



CHANGES
History
=======


0.4 (unreleased)
::::::::::::::::

- Nothing yet.


0.3.1 (2009-04-28)
::::::::::::::::::

- Fix for non-updating index pages despite ALWAYS_REFRESH setting. Cause:
  already downloaded urls were not re-processed.  [reinout]


0.3 (2009-03-16)
::::::::::::::::

- Updated documentation for buildout change and always_refresh change.
  [reinout]

- Attempt ../etc/eggproxy.conf as config location to help buildout setups.
  [reinout]

- Added always_refresh (and timeout) option to always attempt a real pypi
  connection to ensure freshness.  [reinout]

- Added port option to configuration. [reinout]

- don't lowercase package names

- improve wsgi app

- add tests


0.2.0 (2008-09-22)
::::::::::::::::::

- initial version as collective.eggproxy. This package was formerly known as
  iw.eggproxy (https://ingeniweb.svn.sourceforge.net/svnroot/ingeniweb/iw.eggproxy)

- package index/download files: skip modules installed in local system (resulted
  in copying a directory instead of downloading a file)

- fixed Update script crashing with invalid/obsolete package name

- Overriden PackageIndex so that we get eggs distributions for all versions, all
  platforms

- fixed malformed tag in generated indexes

0.1.0 (2008-06-06)
::::::::::::::::::

- initial version created by IngeniSkel


collective/eggproxy/tests/__init__.py
from nose import with_setup
import tempfile
import shutil
import os

tempdir = os.path.join(tempfile.gettempdir(), 'eggs')

def setup_func():
    if os.path.isdir(tempdir):
        shutil.rmtree(tempdir)
    os.mkdir(tempdir)

def teardown_func():
    if os.path.isdir(tempdir):
        shutil.rmtree(tempdir)




collective/eggproxy/tests/test_wsgi.py
# -*- coding: utf-8 -*-
from collective.eggproxy.wsgi import EggProxyApp
from collective.eggproxy.tests import *
from paste.fixture import TestApp

app = TestApp(EggProxyApp('http://pypi.python.org/simple', eggs_dir=tempdir))

@with_setup(setup_func, teardown_func)
def test_root_index():
    response = app.get('/')
    assert 'collective.eggproxy' in response, response

@with_setup(setup_func, teardown_func)
def test_package_index():
    response = app.get('/collective.eggproxy')
    assert '<title>collective.eggproxy</title>' in response, response

@with_setup(setup_func, teardown_func)
def test_package():
    # trailing '/': without this, paste.fixture.TestResponse click results in:
    # >>> urlparse.urljoin('/collective.eggproxy', 'collective.eggproxy-0.2.0.tar.gz'
    # '/collective.eggproxy-0.2.0.tar.gz'
    # actual url is: '/collective.eggproxy/collective.eggproxy-0.2.0.tar.gz'
    response = app.get('/collective.eggproxy/')
    response = response.click(index=0, verbose=True)
    assert ('Content-Encoding', 'gzip') in response.headers, response.headers



collective/eggproxy/tests/test_utils.py
# -*- coding: utf-8 -*-
from collective.eggproxy.utils import IndexProxy
from collective.eggproxy.tests import *

index = IndexProxy()

@with_setup(setup_func, teardown_func)
def test_index():
    # assume non-existence
    assert os.path.isfile(os.path.join(tempdir, 'index.html')) == False

    index.updateBaseIndex(tempdir)

    # assume existence
    assert os.path.isfile(os.path.join(tempdir, 'index.html')) == True

    page = open(os.path.join(tempdir, 'index.html')).read()

    # check case
    assert 'PasteDeploy' in page

@with_setup(setup_func, teardown_func)
def test_package():
    index.updatePackageIndex('collective.eggproxy', tempdir)
    dirname = os.path.join(tempdir, 'collective.eggproxy')

    assert os.path.isdir(dirname) == True

    assert os.listdir(dirname) == ['index.html']
    ['index.html']

    page = open(os.path.join(dirname, 'index.html')).read()
    assert '<html><head><title>collective.eggproxy</title></head>' in page

@with_setup(setup_func, teardown_func)
def test_package_case_sensitive():
    index.updatePackageIndex('Paste', tempdir)
    dirname = os.path.join(tempdir, 'Paste')

    assert os.path.isdir(dirname) == True

    assert os.listdir(dirname) == ['index.html']
    ['index.html']

    page = open(os.path.join(dirname, 'index.html')).read()
    assert '<html><head><title>Paste</title></head>' in page


bootstrap.py
##############################################################################
#
# Copyright (c) 2006 Zope Corporation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""Bootstrap a buildout-based project

Simply run this script in a directory containing a buildout.cfg.
The script accepts buildout command-line options, so you can
use the -c option to specify an alternate configuration file.

$Id$
"""

import os, shutil, sys, tempfile, urllib2

tmpeggs = tempfile.mkdtemp()

ez = {}
exec urllib2.urlopen('http://peak.telecommunity.com/dist/ez_setup.py'
                     ).read() in ez
ez['use_setuptools'](to_dir=tmpeggs, download_delay=0)

import pkg_resources

cmd = 'from setuptools.command.easy_install import main; main()'
if sys.platform == 'win32':
    cmd = '"%s"' % cmd # work around spawn lamosity on windows

ws = pkg_resources.working_set
assert os.spawnle(
    os.P_WAIT, sys.executable, sys.executable,
    '-c', cmd, '-mqNxd', tmpeggs, 'zc.buildout',
    dict(os.environ,
         PYTHONPATH=
         ws.find(pkg_resources.Requirement.parse('setuptools')).location
         ),
    ) == 0

ws.add_entry(tmpeggs)
ws.require('zc.buildout')
import zc.buildout.buildout
zc.buildout.buildout.main(sys.argv[1:] + ['bootstrap'])
shutil.rmtree(tmpeggs)



collective/eggproxy/wsgi.py
# -*- coding: utf-8 -*-
## Product description
##
## Copyright (C) 2008 Ingeniweb

## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2 of the License, or
## (at your option) any later version.

## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.

## You should have received a copy of the GNU General Public License
## along with this program; see the file COPYING. If not, write to the
## Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
import os
import socket
import sys
from paste import httpserver
from paste.script.appinstall import Installer as BaseInstaller
from paste.fileapp import FileApp
from paste.httpexceptions import HTTPNotFound
from collective.eggproxy.utils import PackageIndex
from collective.eggproxy import IndexProxy
from collective.eggproxy import PackageNotFound
from collective.eggproxy.config import config

ALWAYS_REFRESH = config.getboolean('default', 'always_refresh')
if ALWAYS_REFRESH:
    print "Always-refresh mode switched on"
    # Apply timeout setting right here. Might not be the best spot. Timeout is
    # needed for the always_refresh option to keep a down pypi from blocking
    # the proxy.
    timeout = config.get('default', 'timeout')
    socket.setdefaulttimeout(int(timeout))


class EggProxyApp(object):

    def __init__(self, index_url=None, eggs_dir=None):
        if not index_url:
            index_url = config.get('default', 'index')
        if not eggs_dir:
            eggs_dir = config.get('default', 'eggs_directory')
        if not os.path.isdir(eggs_dir):
            print 'You must create the %r directory' % eggs_dir
            sys.exit()
        self.eggs_index_proxy = IndexProxy(PackageIndex(index_url=index_url))
        self.eggs_dir = eggs_dir

    def __call__(self, environ, start_response):
        path = [part for part in environ.get('PATH_INFO', '').split('/')
                if part]
        if len(path) > 2:
            raise ValueError, "too many components in url"

        if len(path) > 0 and path[-1] == 'index.html':
            path.pop()

        path_len = len(path)

        if path_len == 0:
            filename = self.checkBaseIndex()
        elif path_len == 1:
            if path[0] == "favicon.ico":
                return HTTPNotFound()(environ, start_response)
            filename = self.checkPackageIndex(path[0])
        else:
            filename = self.checkEggFor(path[0], path[1])

        if filename is None:
            return HTTPNotFound()(environ, start_response)

        return FileApp(filename)(environ, start_response)

    def checkBaseIndex(self):
        filename = os.path.join(self.eggs_dir, 'index.html')
        if not os.path.exists(filename):
            self.eggs_index_proxy.updateBaseIndex(self.eggs_dir)
        return filename

    def checkPackageIndex(self, package_name):
        filename = os.path.join(self.eggs_dir, package_name, 'index.html')
        if not os.path.exists(filename):
            try:
                self.eggs_index_proxy.updatePackageIndex(
                    package_name,
                    eggs_dir=self.eggs_dir)
            except PackageNotFound:
                return None
        elif ALWAYS_REFRESH:
            # Force refresh
            try:
                self.eggs_index_proxy.updatePackageIndex(
                    package_name,
                    eggs_dir=self.eggs_dir)
            except PackageNotFound:
                pass
        else:
            # Just use the proxied copy.
            pass
        return filename

    def checkEggFor(self, package_name, eggname):
        filename = os.path.join(self.eggs_dir, package_name, eggname)
        if not os.path.exists(filename):
            try:
                self.eggs_index_proxy.updateEggFor(package_name, eggname,
                                                   eggs_dir=self.eggs_dir)
            except ValueError:
                return None
        return filename


def app_factory(global_config, **local_conf):
    # Grab config from wsgi .ini file. If not specified, config.py's values
    # take over.
    eggs_dir = local_conf.get('eggs_directory', None)
    index_url = local_conf.get('index', None)
    return EggProxyApp(index_url, eggs_dir)


class Installer(BaseInstaller):
    use_cheetah = False
    config_file = 'deployment.ini_tmpl'

    def config_content(self, command, vars):
        import pkg_resources
        module = 'collective.eggproxy'
        if pkg_resources.resource_exists(module, self.config_file):
            return self.template_renderer(
                pkg_resources.resource_string(module, self.config_file),
                vars,
                filename=self.config_file)


def standalone():
    port = config.get('default', 'port')
    # 0.2.0 way of starting the httpserver, but using the config'ed port
    # number instead of a hardcoded 8888.
    httpserver.serve(EggProxyApp(), host='127.0.0.1', port=port)
    # Post-0.2.0 way of starting the server using hardcoded config by means of
    # the package-internal .ini file. This does not allow starting it on a
    # different port, so I [reinout] commented it out for now.
    #import paste.script.command
    #this_dir = os.path.dirname(__file__)
    #sys.argv.extend(['serve', os.path.join(this_dir, 'wsgi.ini')])
    #paste.script.command.run()


collective/eggproxy/utils.py
# -*- coding: utf-8 -*-
## Copyright (C) 2008 Ingeniweb

## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2 of the License, or
## (at your option) any later version.

## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.

## You should have received a copy of the GNU General Public License
## along with this program; see the file COPYING. If not, write to the
## Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
import os
import tempfile
import urllib2
import urlparse
import shutil
from setuptools.package_index import PackageIndex as BasePackageIndex
from setuptools.package_index import (
    egg_info_for_url,
    safe_name,
    safe_version,
    to_filename,
    HREF,
    htmldecode,
    find_external_links,
    PYPI_MD5,
    )

from pkg_resources import Requirement
from collective.eggproxy.config import config

ALWAYS_REFRESH = config.getboolean('default', 'always_refresh')
EGGS_DIR = config.get("default", "eggs_directory")
INDEX_URL = config.get("default", "index")
#INDEX is defined *after* the PackageIndex class.


class PackageIndex(BasePackageIndex):
    """
    """

    def can_add(self, dist):
        """Overrides PackageIndex.can_add method to remove filter on python
    major version: we want packages for all versions, all platforms
        """
        return True

    def process_index(self, url, page):
        """Process the contents of a PyPI page
        Override: don't lowercase package name
        """
        if ALWAYS_REFRESH:
            # Zap ourselves from the fetched url list, otherwise we'll
            # never be updated as long as the server runs.
            del self.fetched_urls[url]

        def scan(link):
            # Process a URL to see if it's for a package page
            if link.startswith(self.index_url):
                parts = map(
                    urllib2.unquote, link[len(self.index_url):].split('/')
                )
                if len(parts)==2 and '#' not in parts[1]:
                    # it's a package page, sanitize and index it
                    pkg = safe_name(parts[0])
                    ver = safe_version(parts[1])
                    # changed "pkg.lower()" to "pkg"
                    self.package_pages.setdefault(pkg, {})[link] = True
                    return to_filename(pkg), to_filename(ver)
            return None, None

        # process an index page into the package-page index
        for match in HREF.finditer(page):
            scan( urlparse.urljoin(url, htmldecode(match.group(1))) )

        pkg, ver = scan(url)   # ensure this page is in the page index
        if pkg:
            # process individual package page
            for new_url in find_external_links(url, page):
                # Process the found URL
                base, frag = egg_info_for_url(new_url)
                if base.endswith('.py') and not frag:
                    if ver:
                        new_url+='#egg=%s-%s' % (pkg,ver)
                    else:
                        self.need_version_info(url)
                self.scan_url(new_url)

            return PYPI_MD5.sub(
                lambda m: '<a href="%s#md5=%s">%s</a>' % m.group(1,3,2), page
            )
        else:
            return ""   # no sense double-scanning non-package pages


INDEX = PackageIndex(index_url=INDEX_URL)


class PackageNotFound(Exception):
    """
    """


class IndexProxy(object):

    def __init__(self, index=None):
        self.index = index or INDEX

    def updateBaseIndex(self, eggs_dir=EGGS_DIR):
        """Update base index.html
        """
        file_path = os.path.join(eggs_dir, 'index.html')
        html = open(file_path, 'w')

        self.index.scan_all()
        package_names = self.index.package_pages.keys()
        package_names.sort()

        print >> html, "<html><head><title>Simple Index</title></head><body>"
        for pack_name in package_names:
            print >> html, '<a href="%s/">%s</a><br/>' % (pack_name, pack_name)
        print >> html, '</body></html>'

        html.close()
        del html

    def _lookupPackage(self, package_name):
        requirement = Requirement.parse(package_name)
        self.index.find_packages(requirement)

    def updatePackageIndex(self, package_name, eggs_dir=EGGS_DIR):
        """Update info for a specific package
        """
        self._lookupPackage(package_name)
        if not self.index[package_name]:
            raise PackageNotFound, "Package '%s' does not exists or has no eggs" % package_name
        package_path = os.path.join(eggs_dir, package_name)
        if not os.path.exists(package_path):
            os.mkdir(package_path)

        html_path = os.path.join(package_path, 'index.html')
        dists = self.index[package_name]
        if not dists and os.path.exists(html_path):
            # We already have a cached index page and there are no dists.
            # Pypi is probably down, so we keep our existing one.
            return
        html = open(html_path, 'w')
        title = "Links for %s" % package_name
        print >> html, "<html><head><title>%s</title></head>" % package_name
        print >> html, "<body><h1>%s</h1>" % package_name
        for dist in dists:
            if getattr(dist, "module_path", None) is not None:
                # this is a module installed in system
                continue

            filename, md5 = egg_info_for_url(dist.location)
            print >> html, (
                '<a href="%s#%s" rel="download">%s</a><br />'
                % (filename, md5, filename)
                )

        print >> html, "</body></html>"
        html.close()
        del html

    def updateEggFor(self, package_name, eggname, eggs_dir=EGGS_DIR):
        """Download an egg for package_name
        """
        self._lookupPackage(package_name)
        file_path = os.path.join(eggs_dir, package_name, eggname)
        for dist in self.index[package_name]:
            if getattr(dist, "module_path", None) is not None:
                # this is a module installed in system (directory), we want
                # to download a fresh package
                continue

            filename, md5 = egg_info_for_url(dist.location)
            if filename == eggname:
                tmp = tempfile.gettempdir()
                tmp_location = self.index.download(dist.location, tmp)
                shutil.move(tmp_location, file_path)
                return

        raise ValueError, "Egg '%s' not found in index" % eggname


collective/eggproxy/__init__.py
# -*- coding: utf-8 -*-
## Copyright (C) 2008 Ingeniweb

## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2 of the License, or
## (at your option) any later version.

## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.

## You should have received a copy of the GNU General Public License
## along with this program; see the file COPYING. If not, write to the
## Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
from collective.eggproxy.utils import PackageNotFound
from collective.eggproxy.utils import IndexProxy
eggs_index_proxy = IndexProxy()


collective/eggproxy/update_script.py
# -*- coding: utf-8 -*-
## Copyright (C) 2008 Ingeniweb

## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2 of the License, or
## (at your option) any later version.

## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.

## You should have received a copy of the GNU General Public License
## along with this program; see the file COPYING. If not, write to the
## Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
import os
from os.path import getmtime
from time import time

from collective.eggproxy import eggs_index_proxy
from collective.eggproxy import PackageNotFound
from collective.eggproxy.config import config

UPDATE_INTERVAL = int(config.get("default", "update_interval")) * 3600
EGGS_DIR = config.get("default", "eggs_directory")
TIME_LIMIT = int(time()) - UPDATE_INTERVAL

def isOutDated(file_path):
    """A file is outdated if it does not exists or if its modification date is
    older than now - update_interval
    """
    if os.path.exists(file_path):
        mtime = getmtime(file_path)
        return mtime < TIME_LIMIT
    return True

def updateCache(*args):
    """
    """
    isDir = os.path.isdir
    pathJoin = os.path.join
    update = False
    index_file = pathJoin(EGGS_DIR, 'index.html')

    if isOutDated(index_file):
        eggs_index_proxy.updateBaseIndex()

    for package_name in os.listdir(EGGS_DIR):
        dir_path = pathJoin(EGGS_DIR, package_name)
        if not isDir(dir_path):
            continue

        index_file = pathJoin(dir_path, 'index.html')
        if isOutDated(index_file):
            try:
                eggs_index_proxy.updatePackageIndex(package_name)
            except PackageNotFound, msg:
                # FIXME: use logging
                print msg


collective/eggproxy/apache_handler.py
# -*- coding: utf-8 -*-
## Copyright (C) 2008 Ingeniweb

## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2 of the License, or
## (at your option) any later version.

## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.

## You should have received a copy of the GNU General Public License
## along with this program; see the file COPYING. If not, write to the
## Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
import os
from mod_python import apache

from collective.eggproxy import eggs_index_proxy
from collective.eggproxy.config import config

EGGS_DIR = config.get("default", "eggs_directory")


def fixup_handler(req):
    options = req.get_options()
    document_root = EGGS_DIR
    url_prefix = options['URLPrefix']
    url_prefix = [part for part in url_prefix.split('/') if part]
    url_prefix_len = len(url_prefix)
    uri_path = req.parsed_uri[apache.URI_PATH]
    uri_path = [part for part in uri_path.split('/') if part][url_prefix_len:]

    real_path = os.path.join(document_root, *uri_path)

    if os.path.exists(real_path) and not os.path.isdir(real_path):
        #FIXME isdir: cannot make apache serve index.html with mod_dir
        req.handler = 'default-handler'
        return apache.DECLINED

    req.handler = 'mod_python'
    req.add_handler('PythonHandler', modpython_handler)
    return apache.OK

def modpython_handler(req):

    options = req.get_options()
    document_root = EGGS_DIR
    url_prefix = options['URLPrefix']
    url_prefix = [part for part in url_prefix.split('/') if part]
    url_prefix_len = len(url_prefix)
    uri_path = req.parsed_uri[apache.URI_PATH]
    uri_path = [part for part in uri_path.split('/') if part][url_prefix_len:]

    if len(uri_path) > 0 and uri_path[-1] == 'index.html':
        uri_path.pop()

    real_path = os.path.join(document_root, *uri_path)
    real_path_exists = os.path.exists(real_path)

    if real_path_exists:
        if not os.path.isdir(real_path):
            return apache.DECLINED

        #FIXME we are doing the job of mod_dir
        real_path = os.path.join(real_path, 'index.html')
        if os.path.exists(real_path):
            req.internal_redirect(req.uri + 'index.html')
##             req.content_type = 'text/html'
##             req.sendfile(real_path)
            return apache.OK

    uri_path_len = len(uri_path)

    if uri_path_len > 2:
        # we are not supposed to go deeper than package/filename.egg
        return apache.HTTP_NOT_FOUND

    if uri_path_len == 0:
        # build the base index
        eggs_index_proxy.updateBaseIndex()
        req.internal_redirect(req.uri + 'index.html')
        return apache.OK

    package_name = uri_path[0]

    if uri_path_len == 1:
        try:
            eggs_index_proxy.updatePackageIndex(package_name)
        except ValueError:
            return apache.HTTP_NOT_FOUND

        req.internal_redirect(req.uri)
        return apache.OK
    else:
        eggname = uri_path[1]
        try:
            eggs_index_proxy.updateEggFor(package_name, eggname)
        except ValueError:
            return apache.HTTP_NOT_FOUND

        req.internal_redirect(req.uri)
        return apache.OK

    return apache.HTTP_NOT_FOUND


collective/eggproxy/config.py
# -*- coding: utf-8 -*-
## Copyright (C) 2008 Ingeniweb

## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2 of the License, or
## (at your option) any later version.

## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.

## You should have received a copy of the GNU General Public License
## along with this program; see the file COPYING. If not, write to the
## Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
import os
import sys
from ConfigParser import ConfigParser


# First try: User-specific config file
CONFIG_FILE = os.path.join(os.path.expanduser('~'), 'eggproxy.conf')
# Second try: buildout setup, we're in bin/, config is in etc/.
if not os.path.exists(CONFIG_FILE):
    bin_dir = os.path.dirname(os.path.abspath(sys.argv[0]))
    CONFIG_FILE = os.path.join(bin_dir, '../etc/eggproxy.conf')
# Last try: Global config file in /etc
if not os.path.exists(CONFIG_FILE):
    CONFIG_FILE = '/etc/eggproxy.conf'

#print "Using config file", CONFIG_FILE
config = ConfigParser()
config.add_section("default")
config.set("default", "eggs_directory", "/var/www")
config.set("default", "index", 'http://pypi.python.org/simple')
config.set("default", "update_interval", '24')
config.set("default", "port", '8888')
config.set("default", "always_refresh", '0')
config.set("default", "timeout", '3')

if os.path.exists(CONFIG_FILE):
    config.readfp(open(CONFIG_FILE))
